Update foreach APIs to use scalar lists

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51060 Refactor foreach pointwise ops tests to use OpInfo
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo
* #49712 Refactor foreach unary ops tests to use OpInfo
* #49250 Update foreach_div and foreach_sub logic
* #49249 Update foreach binary ops with a scalar list
* #51131 Refactor ForeachUtils.h
* #49248 Refactor ForeachUnaryOp.cu
* **#48223 Update foreach APIs to use scalar lists**

Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.

Commit:
Update foreach APIs to use scalar lists

[ghstack-poisoned]

Commit:
Update on "[WIP] Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

[ghstack-poisoned]

Commit:
Update on "[WIP] Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

[ghstack-poisoned]

Commit:
Update on "[WIP] Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -137,7 +137,7 @@ bool can_use_fast_route(TensorList tensors, Scalar scalar) {
 }
 
-bool can_use_fast_route(TensorList tensors, ArrayRef<double> scalars) {
+bool can_use_fast_route(TensorList tensors, ArrayRef<Scalar> scalars) {
```
Does this need to check if type promotion will happen between `tensors` and `scalars`?

izdeby replied:
Yes. Its [done](https://github.com/pytorch/pytorch/pull/48224/files?file-filters%5B%5D=.h&file-filters%5B%5D=.py#diff-0e978890cde97eede8219604e17b10129dae0f8317a8c14272c39d62b874c831R213) in the next PR of this stack.

izdeby replied:
Hmmm, should i move these logic from the following PR to this one?

zou3519 replied:
Hmm... is it possible to split up the next PR in the stack? I haven't actually looked at it yet, but each of the bullet points in the "What was done" section sound standalone

My 2 cents:
- each commit on the stack should stand by itself. If we have to revert the next PR in the stack for some reason, or if this PR gets merged before the next PR, then we are looking at potentially buggy behavior in the intermediate state.
- The next PR in the stack is a whopping +1,835 −930 line diff. The delta is +900. That's pretty large, folks generally have a hard time reviewing anything that's over 300 lines lol.

izdeby replied:
Done. Broke it up into several smaller PRs. Each has its logical idea and leads to next PR.

zou3519 replied:
Thank you!

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -49,7 +49,7 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
 }
 
-void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, TensorList tensors3, ArrayRef<double> scalars) {
+void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, TensorList tensors3, ArrayRef<Scalar> scalars) {
```
Does this need to check if type promotion will happen between tensors and scalars?

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7146,63 +7146,63 @@
     CUDA: foreach_tensor_div_list_kernel_cuda_
 
-- func: _foreach_add.ScalarList(Tensor[] tensors, float[] scalars) -> Tensor[]
+- func: _foreach_add.ScalarList(Tensor[] tensors, Scalar[] scalars) -> Tensor[]
```
I have a question about the PythonArgParser. Given an integer list in python, like [0, 1, 2], which of the following do we parse it as?
- IntList
- FloatList
- ScalarList
Also, now that we have ScalarList, do we still need the codegen for float[] ? This might be a better question for the other diff in this stack

zou3519 replied:
Hmm, this will only be ambiguous if there are overloads of an operator that use some combination of the above (e.g., IntList overload, ScalarList overload)

zou3519 commented on test/test_foreach.py:
```diff
@@ -599,11 +604,42 @@ def test_complex_scalarlist(self, device, dtype):
-                with self.assertRaisesRegex(TypeError, "argument 'scalars' must be tuple of floats"):
+                if self.device_type == 'cpu':
                     res = foreach_bin_op(tensors, scalars)
+                    self.assertEqual(res, expected)
 
-                with self.assertRaisesRegex(TypeError, "argument 'scalars' must be tuple of floats"):
-                    foreach_bin_op_(tensors, scalars)
+                    if dtype not in [torch.complex64, torch.complex128]:
+                        with self.assertRaisesRegex(RuntimeError, "can't be cast to the desired output type"):
+                            foreach_bin_op_(tensors, scalars)
+                    else:
+                        self.assertEqual(res, expected)
+                else:
+                    # float types
+                    if dtype in torch.testing.get_all_fp_dtypes(include_half=True, include_bfloat16=True):
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op_(tensors, scalars)
+
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op(tensors, scalars)
+                        return
+
+                    # complex types
+                    if dtype in [torch.complex64, torch.complex128]:
+                        with self.assertRaisesRegex(RuntimeError, "not implemented for"):
+                            foreach_bin_op(tensors, scalars)
+
+                        with self.assertRaisesRegex(RuntimeError, "not implemented for"):
+                            foreach_bin_op_(tensors, scalars)
+                        return
+
+                    # int types
+                    if dtype in torch.testing.integral_types():
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op_(tensors, scalars)
+
+                        with self.assertRaisesRegex(RuntimeError, "value cannot be converted to type"):
+                            foreach_bin_op(tensors, scalars)
+                        return
```
Some things I am concerned about:
- Why does the complex behavior differ on CPU/CUDA? Do we want those to be ultimately the same?
- The integral cuda and the float cuda logic the same to me, should we just combine them?

izdeby replied:
Please, see the next PR in this stack. I've rewrote the tests to have a separate test per operation. For example, this is how this tests looks [there](https://github.com/pytorch/pytorch/blob/87ce1f9f10b3f694987ff4aea07b12f08effb67c/test/test_foreach.py#L1464).

Much cleaner and easier to follow.

zou3519 commented on test/test_foreach.py:
```diff
@@ -698,57 +738,44 @@ def test_bool_scalarlist(self, device, dtype):
+                                    self.assertEqual(res, expected)
+
+                                foreach_bin_op_(tensors, scalars)
+                                self.assertEqual(res, tensors)
```
Some general feedback is: the tests are really hard to read because they branch so much. I'm not sure how to make them easier to read, though. Perhaps we should consider splitting them up or think about why the behavior diverges so much?

izdeby replied:
Yes, i fully agree. thats why in the next PR in this stack i separate all the tests. Much easier to read and maintain.

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "[WIP] Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "[WIP DONT REVIEW]Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -87,19 +87,25 @@ bool has_same_attributes(Device expected_device, TensorList tensors) {
 bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
   // complex scalar + integral or boolean tensor will result in complex tensor
   if (scalar.isComplex() && at::isIntegralType(tensor.scalar_type(), /*includeBool*/ true)) {
-    return false;
+    return true;
+  }
+
+  // complex scalar + float tensor will result in complex tensor
+  if (scalar.isComplex() && at::isFloatingType(tensor.scalar_type())) {
+    return true;
```
Can we reuse at::result_type and just do (pseudocode):
```
bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
    auto dtype = tensor.dtype();
    auto result_dtype = at::result_type(tensor, scalar);
    return dtype != result_dtype;
}
```
?
This makes it so that we don't need to maintain our separate implementation of this. I don't remember if there was some reason why we chose not to do this originally.

izdeby replied:
Updated

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -137,8 +143,18 @@ bool can_use_fast_route(TensorList tensors, Scalar scalar) {
 
-bool can_use_fast_route(TensorList tensors, ArrayRef<double> scalars) {
-  return can_use_fast_route(tensors);
+bool can_use_fast_route(TensorList tensors, ArrayRef<Scalar> scalars) {
+#ifdef __HIP_PLATFORM_HCC__
+  return false;
+#else
+  for (int i = 0; i < tensors.size(); i++) {
+    if (will_promote_tensor(tensors[i], scalars[i])) {
+      return false;
+    }
+  }
+
+  return true;
+#endif
```
Do we need to use has_same_attributes on `tensors`? E.g. to check that all tensors have the same device and are all strided?

izdeby replied:
For all the APIs, im calling `has_same_attributes` first and then `can_use_fast_route`. The reason for that is that these two APIs are checking for different thing so i didn't want to introduce confusion.

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -13,7 +13,7 @@ void check_foreach_api_restrictions(TensorList tensors) {
 }
 
-void check_foreach_api_restrictions(TensorList tensors, ArrayRef<double> scalars) {
+void check_foreach_api_restrictions(TensorList tensors, ArrayRef<Scalar> scalars) {
```
I'm slightly confused about what the difference is between `check_foreach_api_restrictions` and `can_use_fast_path`. It would be nice to have some comments in the code that delineate what each is responsible for.

From my understanding:

check_foreach_api_restrictions
- Tensor lists must be non-empty.
- All tensors in all lists must have the same dtype.
- All TensorLists and ScalarLists must have the same number of elements.
- Corresponding tensors must have the same size.

`can_use_fast_route`
- all tensors must have the same device
- all tensors must be non-overlapping and strided
- corresponding tensors must have the same strides
- type promotion must not occur (and the only time this can occur is TensorList-ScalarList operations)

Is that correct? am I missing anything?

izdeby replied:
`check_foreach_api_restrictions` checks API inputs. If restrictions are not met - error is thrown. 
`can_use_fast_path` checks if we can go via 'fast' path. if not, go via slow path. no errors. 

You are correct and i dont think you are missing anything. Im planning to add an additional PR with code docs only but once next few PRs will land. The reason for that is that this logic will change a lot. 

--
Added a small comment around `check_foreach_api_restrictions`. `can_use_fast_path` already has one.

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

zou3519 commented on test/test_foreach.py:
```diff
@@ -142,7 +128,7 @@ def _test_pointwise_op(self, device, dtype, foreach_op, foreach_op_, torch_op):
 
     def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_op):
-        for N in [30, 300]:
+        for N in N_values:
```
nit: global constants should be all caps ("N_VALUES")

zou3519 commented:
Hmm, there are some failing tests. I don't know if they're related (run with CUDA_LAUNCH_BLOCKING=1 locally to check: https://app.circleci.com/pipelines/github/pytorch/pytorch/251834/workflows/52d554fa-945c-47e3-8894-14a1ec435f35/jobs/9671462)

ngimel commented:
They definitely look related.

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

izdeby committed:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

Commit:
Update on "Update foreach APIs to use scalar lists"


Differential Revision: [D25074763](https://our.internmc.facebook.com/intern/diff/D25074763)

**Motivation**
Update existing _foreach APIs to use ScalarList instead of at::ArrayRef<double>

**Testing**
Update the tests assuming that any scalar type can be passed now, not just double.


[ghstack-poisoned]

vincentqb commented:
I see errors on master after this pull request landed. Can we confirm the errors are not related to this PR?
- [pytorch_linux_xenial_cuda10_2_cudnn7_py3_gcc7_test2](https://app.circleci.com/pipelines/github/pytorch/pytorch/268562/workflows/0778c612-56f5-4626-ac22-60c1da328d28/jobs/10628695)
- [pytorch_linux_xenial_cuda11_1_cudnn8_py3_gcc7_test](https://app.circleci.com/pipelines/github/pytorch/pytorch/268562/workflows/0778c612-56f5-4626-ac22-60c1da328d28/jobs/10628694)
- [pytorch_linux_xenial_cuda9_2_cudnn7_py3_gcc7_test](https://app.circleci.com/pipelines/github/pytorch/pytorch/268562/workflows/0778c612-56f5-4626-ac22-60c1da328d28/jobs/10628692)
- [pytorch_windows_vs2019_py36_cpu_test1](https://app.circleci.com/pipelines/github/pytorch/pytorch/268562/workflows/dbbcb2e6-26e8-4e0a-8fd5-12eddaa8e7cc/jobs/10621290)

ngimel commented:
I've reverted this PR, there's IMA failure in this PR's CI, and master is currently broken.
